router/diag_cntl.c

```diff
@@ -205,6 +205,42 @@ struct diag_cntl_cmd_dereg {
 } __packed;
 #define to_cmd_dereg(h) container_of(h, struct diag_cntl_cmd_dereg, hdr)
 
+/* strlcpy is from OpenBSD and not supported by Linux Embedded.
+ * GNU has an equivalent g_strlcpy implementation into glib.
+ * Featurized with compile time USE_GLIB flag for Linux Embedded builds.
+ */
+#ifdef USE_GLIB
+#define strlcpy g_strlcpy
+#endif
+
+#define DIAGID_VERSION_1	1
+#define DIAGID_VERSION_2	2
+#define DIAGID_VERSION_3	3
+
+#define DIAG_CNTL_CMD_DIAG_ID		33
+#define MAX_DIAGID_STR_LEN			30
+#define MIN_DIAGID_STR_LEN			5
+
+struct diag_cntl_cmd_diag_id {
+	struct diag_cntl_hdr hdr;
+	uint32_t version;
+	uint32_t diag_id;
+	char process_name[MAX_DIAGID_STR_LEN];
+} __packed;
+#define to_cmd_diagid(h) container_of(h, struct diag_cntl_cmd_diag_id, hdr)
+
+struct diag_cntl_cmd_diag_id_v2 {
+	struct diag_cntl_hdr hdr;
+	uint32_t version;
+	uint32_t diag_id;
+	uint32_t feature_len;
+	uint32_t pd_feature_mask;
```

Comment:
This seems a bit misleading since the definition of this packet is that feature_mask here is supposed to be variable length size based on feature_len. We should find a better a way to describe this control packet

Reply:
These two fields "feature_len/pd_feature_mask" are not necessary, I just removed them

Reply:
I‘ve been thinking a lot about this, and it might be more appropriable to use a pointer to indicate that the "mask" data length is variable. So I tried to use a pointer to describe the "feature_mask" in this version.